contracts/finance/andromeda-weighted-distribution-splitter/src/contract.rs

```diff
@@ -0,0 +1,938 @@
+    }
+}
+
+pub fn execute_add_recipient(
+    deps: DepsMut,
+    info: MessageInfo,
+    recipient: AddressWeight,
+) -> Result<Response, ContractError> {
+    // Only the contract's owner can add a recipient
+    require(
+        ADOContract::default().is_contract_owner(deps.storage, info.sender.as_str())?,
+        ContractError::Unauthorized {},
+    )?;
+    // No need to send funds
+    require(
+        info.funds.is_empty(),
+        ContractError::FunctionDeclinesFunds {},
+    )?;
+    // Check if splitter is locked
+    let mut splitter = SPLITTER.load(deps.storage)?;
+
+    require(!splitter.locked, ContractError::ContractLocked {})?;
+
+    // Check for duplicate recipients
+
+    let user_index = splitter
+        .recipients
+        .clone()
+        .into_iter()
+        .position(|x| x.recipient == recipient.recipient);
+
+    // If the index exists, the recipient is a duplicate
+    // If the index doesn't exist, it's a new recipient and we may proceed
+    require(user_index == None, ContractError::DuplicateRecipient {})?;
+
+    splitter.recipients.push(recipient);
+    let new_splitter = Splitter {
+        recipients: splitter.recipients,
+        locked: splitter.locked,
+    };
+    SPLITTER.save(deps.storage, &new_splitter)?;
+
+    Ok(Response::default().add_attributes(vec![attr("action", "added_recipient")]))
+}
+
+pub fn execute_andromeda(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: AndromedaMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        AndromedaMsg::Receive(..) => execute_send(deps, info),
+        _ => ADOContract::default().execute(deps, env, info, msg, execute),
+    }
+}
+
+fn execute_send(deps: DepsMut, info: MessageInfo) -> Result<Response, ContractError> {
+    let sent_funds: Vec<Coin> = info.funds.clone();
+    // Amount of coins sent should be at least 1
+    require(
+        !sent_funds.is_empty(),
+        ContractError::InvalidFunds {
+            msg: "Require at least one coin to be sent".to_string(),
+        },
+    )?;
+    // Can't send more than 5 types of coins
+    require(
+        info.funds.len() < 5,
+        ContractError::ExceedsMaxAllowedCoins {},
+    )?;
+
+    let splitter = SPLITTER.load(deps.storage)?;
+    let mut msgs: Vec<SubMsg> = Vec::new();
+
+    let mut remainder_funds = info.funds.clone();
+
+    let mut total_weight = Uint128::zero();
+
+    // Calculate the total weight of all recipients
+    for recipient_addr in &splitter.recipients {
+        let recipient_weight = recipient_addr.weight;
+        total_weight += recipient_weight;
+    }
+
+    // Each recipient recieves the funds * (the recipient's weight / total weight of all recipients)
+    // The remaining funds go to the sender of the function
+    for recipient_addr in &splitter.recipients {
+        let recipient_weight = recipient_addr.weight;
+        let mut vec_coin: Vec<Coin> = Vec::new();
+        for (i, coin) in sent_funds.iter().enumerate() {
```

Comment:
I think `into_iter` makes more sense here since you are just cloning in the loop anyways.